pkg/service/service.go

```diff
@@ -1401,3 +1405,74 @@ func isWildcardAddr(frontend lb.L3n4AddrID) bool {
 	}
 	return net.IPv4zero.Equal(frontend.IP)
 }
+
+// SyncServicesOnDeviceChange finds and adds missing load-balancing entries for
+// new devices.
+func (s *Service) SyncServicesOnDeviceChange(nodeAddressing types.NodeAddressing) {
+	s.Lock()
+	defer s.Unlock()
+
+	existingFEs := make(map[string]struct{})
+
+	// Find all NodePort services by finding the surrogate services.
+	nodePortSvcs := make([]*svcInfo, 0)
+	for _, svc := range s.svcByID {
+		if svc.svcType != lb.SVCTypeNodePort {
+			continue
+		}
+
+		if svc.frontend.IP.IsUnspecified() {
+			nodePortSvcs = append(nodePortSvcs, svc)
+		} else {
+			existingFEs[svc.frontend.IP.String()] = struct{}{}
+		}
+	}
+
+	// Create the missing IPv4 nodeport services
+	if option.Config.EnableIPv4 {
+		for _, ip := range nodeAddressing.IPv4().LoadBalancerNodeAddresses() {
+			if ip.IsUnspecified() {
+				continue
+			}
+			if _, ok := existingFEs[ip.String()]; !ok {
+				// No services for this frontend, create them.
+				for _, svcInfo := range nodePortSvcs {
+					fe := lb.NewL3n4AddrID(svcInfo.frontend.Protocol, ip, svcInfo.frontend.Port, svcInfo.frontend.Scope, 0)
+					svc := svcInfo.deepCopyToLBSVC()
+					svc.Frontend = *fe
+
+					_, _, err := s.upsertService(svc)
+					if err != nil {
+						log.WithError(err).Warnf("Could not create service for frontend %s", ip)
```

Comment:
Please change this to use structured logging, `log.WithFields(logfields.IPAddr, ip).WithError(err).Warn(...)`. This allows logging systems to aggregate messages based on the message field and correlate the same type of warning between nodes etc. because the dynamic content (IP) is held in a different IP. 

(may or may not apply to other logging statements elsewhere, I won't go through all of them).

Reply:
Done.